source/common/http/utility.cc

```diff
@@ -463,9 +463,26 @@ void Utility::appendVia(RequestOrResponseHeaderMap& headers, const std::string&
 
 void Utility::updateAuthority(RequestHeaderMap& headers, absl::string_view hostname,
                               const bool append_xfh) {
-  if (append_xfh && !headers.getHostValue().empty()) {
-    headers.appendForwardedHost(headers.getHostValue(), ",");
+  const auto host = headers.getHostValue();
+
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.append_xfh_idempotent")) {
+    // Only append to x-forwarded-host if the value was not the last value appended.
+    const auto xfh = headers.getForwardedHostValue();
+
+    if (append_xfh && !host.empty() && !xfh.empty()) {
+      const auto xfh_split = StringUtil::splitToken(xfh, ",");
+      if (!xfh_split.empty() && xfh_split.back() != host) {
+        headers.appendForwardedHost(host, ",");
+      }
+    } else if (append_xfh && !host.empty()) {
```

Comment:
nit: these conditions are the same as the first conditions of the previous if statement. This can probably be refactored into:
```
if (append_xfh && !host.empty()) {
  ...
  if (!xfh.empty()) {
    ...
  } else {
    ...
  }
}
```